File: core/BinaryTreeZigzagLevelOrderTraversal/BinaryTreeZigzagLevelOrderTraversal.py
from typing import List
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def zigzagLevelOrder(root: TreeNode) -> List[List[int]]:
    if not root:
        return []

    queue = deque([root])
    res, count = [], 0
    while queue:
        count += 1
        length, temp = len(queue), []
        for _ in range(length):
            node = queue.popleft()
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
            temp.append(node.val)
        if count % 2 == 0:
            temp.reverse()
            res.append(temp)
        else:
            res.append(temp)

    return res

File: core/BinaryTreeZigzagLevelOrderTraversal/BinaryTreeZigzagLevelOrderTraversal.java
package com.dbc;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class BinaryTreeZigzagLevelOrderTraversal {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if (root == null) return new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<TreeNode>() {{
            add(root);
        }};
        List<List<Integer>> res = new ArrayList<>();
        boolean flag = false;
        while (!queue.isEmpty()) {
            int length = queue.size();
            flag = !flag;
            LinkedList<Integer> temp = new LinkedList<>();
            for (int i = 0; i < length; i ++) {
                TreeNode node = queue.poll();
                if (node.left != null) queue.add(node.left);
                if (node.right != null) queue.add(node.right);
                if (!flag) temp.offerFirst(node.val);
                else temp.offerLast(node.val);
            }
            res.add(new LinkedList<>(temp));
        }
        return res;
    }
}
